libzecale/circuits/aggregator_gadget.tcc

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2015-2020 Clearmatics Technologies Ltd
+//
+// SPDX-License-Identifier: LGPL-3.0+
+
+#ifndef __ZECALE_CIRCUITS_AGGREGATOR_GADGET_TCC__
+#define __ZECALE_CIRCUITS_AGGREGATOR_GADGET_TCC__
+
+#include "libzecale/circuits/aggregator_gadget.hpp"
+
+namespace libzecale
+{
+
+template<
+    typename nppT,
+    typename wppT,
+    typename nsnarkT,
+    typename wverifierT,
+    size_t NumProofs>
+aggregator_gadget<nppT, wppT, nsnarkT, wverifierT, NumProofs>::
+    aggregator_gadget(
+        libsnark::protoboard<libff::Fr<wppT>> &pb,
+        const size_t inputs_per_nested_proof,
+        const std::string &annotation_prefix)
+    : libsnark::gadget<libff::Fr<wppT>>(pb, annotation_prefix)
+    , num_inputs_per_nested_proof(inputs_per_nested_proof)
+{
+    // Assert that a single nested input can be encoded in a single wrapper
+    // input. This holds for all pairing chains we test, but if it fails then
+    // nested inputs will need to be encoded over multiple wrapper inputs.
+    // Unfortunately this check cannot be made statically.
+    //
+    // TODO: if num_bits are equal then also check mod values.
+    assert(libff::Fr<nppT>::num_bits <= libff::Fr<wppT>::num_bits);
+
+    // TODO: allocations of these "public" variables should happen outside of
+    // the gadget.
+
+    // Allocate the primary inputs and results first (these are expected to be
+    // public).
+    for (size_t i = 0; i < NumProofs; i++) {
+        nested_primary_inputs[i].allocate(
+            pb,
+            num_inputs_per_nested_proof,
+            FMT(this->annotation_prefix,
+                " nested_primary_inputs_bits[%zu]",
+                i));
+
+        // Allocation of the results
+        nested_proofs_results[i].allocate(
+            pb, FMT(this->annotation_prefix, " nested_proofs_results[%zu]", i));
+    }
+
+    // TODO: Allocate (or accept) a variable to store the hash of the VK (to
+    // avoid proofs generated with malicious keypair).
```

Comment:
Likewise for the "nested instances"

Reply:
I'm not sure I understand the comment.  This TODO is just about (eventually) allocating the primary input for the hask of the nested verification key (which much be done before other varibale are allocated).

Reply:
My comment above is just a remark saying that, yes we do need to hash the VK, but we will also -  eventually - need to hash the instance before calling the nested SNARK verification algo to "collapse" the instance (for arbitrary statements) to the same constant inputs which is necessary to make the Zecale statement generic

Reply:
That's not a big deal, there's already a ticket for that: https://github.com/clearmatics/zecale/issues/5